Updated SRT core statistics documentation (#1199)
Haivision/srt: docs/statistics.md

# SRT Statistics

SRT provides a powerful set of statistical data on a socket. This data can be used to keep an eye on a socket's health and track faulty behavior.

Statistics are calculated independently on each side (receiver and sender) and are not exchanged between peers unless explicitly stated.

> TODO: unless explicitly stated ??? -> check this on msRTT, mbpsBandwidth

Reviewer: The average RTT and Bandwidth measurements are being performed at the receiver side and sent back to sender through UMSG_ACK message. Mention this as exception and you don't need extra TODOs.

## Accumulated Statistics

### msTimeStamp

The time elapsed since the SRT socket has been created (after successful call to `srt_connect(...)` or `srt_bind(...)` function), in milliseconds. Available both for sender and receiver.

### pktSentTotal

The total number of sent data packets, including retransmitted packets. Available for sender.

### pktRecvTotal

The total number of received packets, including retransmitted packets. Available for receiver.

### pktSndLossTotal

The total number of data packets considered or reported as lost at the sender side. Does not correspond to the packets detected as lost at the receiver side. Available for sender.

A packet is considered lost in two cases: 
1. Sender receives a loss report from a receiver,
2. Sender initiates retransmission after not receiving an ACK packet for a certain timeout. Refer to `FASTREXMIT` and `LATEREXMIT` algorithms.

### pktRcvLossTotal

TODO: 

- Update the description.
- Think whether we are going to split lost and reordered packets here and provide user a formula how to get lost only packets if not

The total number of data packets detected as lost at the receiver side. Available for receiver.

> If a packet is received out of order, the gap (sequence discontinuity) is also taken into account and the number of lost packets is increased by the discontinuity size independently of the reordering tolerance value.

Reviewer: I'd hold up this description until it's clarified by fixing #899. The current state is kinda wrong and reordered packets are now counted as lost. By mentioning the number of reordered packets though you gave me an idea. The missing sequences can be first collected as reordered packets (if reorder tolerance is working), and then if they turn out to not arrive when the reorder tolerance elapses, they would be "re-qualified" as lost (that is, number of these packets would be subtracted from reordered stat and added to loss stat).

Author: I created a ticket #1212 for this. I honestly have a lot of concerns regarding this statistics. If you can described in more details how the reorder tolerance is taken into account, it will help. But as I described in the ticket, the reordered packets should be calculated independently on the reorder tolerance. They are simple in-order or out-of-order. Please add the details in the ticket, not here. I'll resolve the current conversation.

Reviewer: Generally you have a handler for every incoming packet. If a jump-over is seen, and your current reorder tolerance is > 0, this gap is notified in the "fresh loss" record for now. After so many packets pass as the reorder tolerance and the packet still didn't arrive, only then is it treated as lost and LOSSREPORT is sent.

The reorder tolerance is a dynamic value, however. It may happen that the packet was reordered, but it was treated as lost because it didn't arrive before the tolerance elapsed (or it's 0 now). But if a belated packet with R=0 is seen, then the distance between its sequence and top receiver sequence is set as the new reorder tolerance (but only up to maximum LOSSMAXTTL value).

The result is that it simply gives it time to receive the reordered packet without making a noise of a lost packet, but with some limits, so it still may happen that some packet will arrive with a greater reorder distance, and therefore will be treated mistakenly as lost.

But then, the method I suggested last time should work. The reorder prevention mechanism must have some limits in which it tolerates the reordering - but statistics don't need it. They can extend the tolerance for packet reception even in infinity - let's say up to the ACK - so that a packet that arrived past the tolerance as reordered, but it was already loss-reported, will still be counted as reordered, not as lost.

> Lost packets detection is based on the gaps in sequence numbers of the SRT DATA packets and is triggered by a newly received packet. An offset is calculated between sequence numbers of the newly arrived DATA packet and the previously received DATA packet (the packet with the highest sequence number). Receiving older packets does not affect this value. The packets from that gap are considered lost,

Reviewer: I'd use here the term of "missing" packets - and they can potentially turn out to be either lost or reordered.

Author: I came to the same idea actually, everything is in #1212.

and that number is added to the `pktRcvLossTotal` measurement. In the case where
the offset is negative, the packet is considered late, meaning that it was either
already acknowledged or dropped by TSBPD as too late to be delivered. Such late

> packets are ignored.

Reviewer: That's kinda non-precise. If we state that we calculate the offset between the newly arriving packet and the top receiver sequence, depending on this offset the packet is qualified this way:
* off = 1 --> this is the very next packet, as expected
* off > 1 --> we have missing packets
* off < 1 --> late packet

This "late packet" by this definition is a packet that could be either
* received out-of-order (R=0)
* retransmitted (R=1)

The "belated" packets (there's a separate stat for that - `pktRcvBelated`) are packets that:
* Are behind the current ACK
* Are ahead of current ACK, but the receiver buffer reported that it already has this packet (that can also happen when an out-of-order packet arrived, while it has been already recovered by FEC)

TODO: 

- This is wrong -> ticket, correct this and make separate statistic. Decrypted statistic should include wrongly encrypted packets. Here we should save the comment regarding the version 1.4.0-1.4.1 and include the comment that since version 1.5.0 we do not take this into account.

> - Calculate loss for FEC -> ticket (low priority), there is something already

Reviewer: FEC filters report packets failed to be rebuilt in `rcvFilterLoss`. These packets can still be recovered in the `arq=onreq` mode, if failed (or if `arq=never`), they will be qualified as dropped.

- We have already pktRcvUndecryptTotal

Since SRT v1.4.0, includes packets that failed to be decrypted.

### pktRetransTotal

The total number of retransmitted packets. Available for sender.

TODO: 

> - We can calculate retransmitted packets both at the sender side and at the receiver side which is stated in pktSentTotal and pktRecvTotal stats -> we should have this statistics available both for snd and rcv. The same for byteRetransTotal -> rename this to pktSentRetransTotal plus add new statistics pktRecvRetransTotal, be careful this is API change. So, think how we can use two names pktRetransTotal and pktSentRetransTotal, until pktRetransTotal can be deprecated. And here we should put special comment. Create ticket for this.

Reviewer: Retransmitted packets calculated on the sender count those that departed the sender; those calculated on the receiver count those that arrived at the receiver, which is less by the number of retransmitted packets lost. Sides do not interchange this information anyhow. For example, if you want to know how many retransmitted packets have been "lost again", you need to process stats collected separately on both sides. The difference between `pktRetrans` on sender and `pktRcvRetrans` on receiver is the number of "lost again" packets. Note that `pktRcvRetrans` is a newly introduced stat since the R retransmission flag has been added - without this flag distinction between unordered and retransmitted packets would not be possible.

- Not exchanged with the receiver: what does it mean?

### pktSentACKTotal

> The total number of sent ACK packets. Available for sender.

Reviewer 2: ```suggestion
The total number of sent ACK packets. Available for receiver.
```

TODO: 

> - possible mistake, see pktSentNAKTotal -> receiver side? -> Maxim should check (Available for sender. -> receiver)

Reviewer: Receiver is the side that sends ACK and LOSSREPORT. That's why this is a stat available at the receiver only.

Conversely, the sender is the side that counts `pktRecvACKTotal` etc.

### pktRecvACKTotal

The total number of received ACK packets. Available for receiver.

> TODO: Possible mistake, see pktSentNAKTotal. -> Maxim to check

Reviewer 2: - [x] Checked. The proposed version "Available for receiver" is correct.

Author: Then I do not understand. ACKACK from sender?

Reviewer: ACKACK is sent in the same direction as data. ACK and LOSSREPORT are sent in the reverse direction.

### pktSentNAKTotal

> The total number of NAK packets received by the sender from the receiver. Essentially means LOSS reports. Available for sender.

Reviewer 2: ```suggestion
The total number of NAK (Negative Acknowledgement) packets sent by the receiver.
Essentially means LOSS reports. Available for receiver.
```

Reviewer: I'd mention more precisely the `UMSG_LOSSREPORT` type message.

TODO: 

- abbreviation
- sent -> received - right? or if it's sent than it's receiver statistics. Correct in the table as well. -> Maxim should check

### pktRecvNAKTotal

> The total number of NAK packets sent back to the sender by the receiver. Essentially means LOSS reports. Available for receiver.

Reviewer 2: ```suggestion
The total number of NAK (Negative ACK) packets received by the sender from the receiver.
It essentially means LOSS reports. Available for the receiver.
```

Author: Available for sender then?

TODO: received -> sent? Or if it's received, than it's sender statistics. Correct in the table as well. -> Maxim to check

### usSndDurationTotal

The total accumulated time in microseconds, during which the SRT sender has some 
data to transmit, including packets that have been sent, but not yet acknowledged. In other words, the total accumulated duration in microseconds when there was something to deliver (non-empty senders' buffer). Available for sender.

### pktSndDropTotal

The total number of "too late to send" packets dropped by the sender (refer to `SRTO_TLPKTDROP` in [API.md](API.md)). Available for sender.

TODO: Update what's below -> Maxim to check -> issue on Github - update RFC and here put the link

The total delay before `TLPKTDROP` is triggered consists of the `SRTO_PEERLATENCY`, 
plus `SRTO_SNDDROPDELAY`, plus 2 * the ACK interval (default ACK interval is 10 ms).
The delay used is the timespan between the very first packet and the latest packet 
in the sender's buffer.

### pktRcvDropTotal

TODO: 

- missing?
- Revise the text

> - create ticket -> rename back SRTO_TSBPDDELAY -> Maxim to create

Reviewer: Since introduction of bidirectional live connection, there are two latency settings, one per direction. The `SRTO_TSBPDDELAY` name comes from the times before it. The negotiated (effective) latency in particular direction is available through `SRTO_PEERLATENCY` on the sender side and `SRTO_RCVLATENCY` on the receiver side. The existing `SRTO_LATENCY` option is deprecated and its management is done through some compromise.

The total number of "too late to deliver" missing packets. Available for receiver.

Missing packets here means lost or not yet received out-of-order packets. The receiver drops only those packets that are missing by the time there is at least one packet ready to be delivered to the upstream application.

Also includes packets that failed to be decrypted (see [pktRcvUndecryptTotal](#pktRcvUndecryptTotal)). These 
packets are present in the receiver's buffer and not dropped at the moment the decryption has failed. 

Note that `SRTO_TSBPDMODE` and `SRTO_TLPKTDROP` socket options should be enabled (refer to API.md - link!!!!).

### pktRcvUndecryptTotal

The total number of packets that failed to be decrypted at the receiver side. Available for receiver.

> TODO: Is it correct that we include this number in the other statistics?

Reviewer: A packet that was failed to be decrypted behaves as a packet that was sent, but not delivered to the application - ergo, as dropped. It's not lost, though - this is a stat for the network.

### pktSndFilterExtraTotal

TODO: 

- Packet filter control packets are SRT DATA packets. -> put this for all related stats
- put this in FEC docs

The total number of packet filter control packets supplied by the packet filter (refer to
[SRT Packet Filtering & FEC](packet-filtering-and-fec.md)). Sender only.

> Packet filter control packets are SRT DATA packets.

Reviewer: Sure, but for statistics this doesn't mean anything. This packet is not calculated as sent data packets, nor is it treated as data packet at the reception side. I'm not sure (need to dig) how this packet is treated in case when it was detected as loss (the information about lost FEC control packets is available during group decommissioning in FEC, but I don't remember if this information is remembered anyhow in the stats).

Reviewer: BTW. it would be worth checking if the FEC control packets are correctly calculated as the number of sent bytes, same as ACKACK packets, which maybe are small, but they do contribute to the overall bandwidth usage. Stating that packets sent and received do not count FEC control packets, the meaning of these fields is slightly changed. Maybe it's right to calculate them separately, but then the byte stats for FEC control packets should be also added.

Author: See issue #1211.

Introduced in SRT v1.4.0.

### pktRcvFilterExtraTotal

TODO: 

- I do not understand the description, check the whole group.

> - for the whole group we should also mention that this statisitcs is available if FEC option is enabled

Reviewer: This is simply a number of received FEC control packets. The description is very general because this stat is calculated in the packet filter framework code, not in the FEC code.

Introduced in SRT v1.4.0.

### byteSentTotal

Same as `pktSentTotal`, but expressed in bytes, including payload and all the headers (20 bytes IPv4 + 8 bytes UDP + 16 bytes SRT). Sender side.

### byteRecvTotal

Same as `pktRecvTotal`, but expressed in bytes, including payload and all the headers (20 bytes IPv4 + 8 bytes UDP + 16 bytes SRT). Receiver side.

### byteRcvLossTotal

Same as `pktRcvLossTotal`, but expressed in bytes, including payload and all the headers (20 bytes IPv4 + 8 bytes UDP + 16 bytes SRT). Receiver side.

### byteRetransTotal

Same as `pktRetransTotal`, but expressed in bytes, including payload and all the headers (20 bytes IPv4 + 8 bytes UDP + 16 bytes SRT). Sender side.

### byteSndDropTotal

Same as `pktSndDropTotal`, but expressed in bytes, including payload and all the headers (20 bytes IPv4 + 8 bytes UDP + 16 bytes SRT). Sender side.

### byteRcvDropTotal

Same as `pktRcvDropTotal`, but expressed in bytes, including payload and all the headers (20 bytes IPv4 + 8 bytes UDP + 16 bytes SRT). Bytes for the dropped packets' payloads are estimated based on the average packet size. Receiver side.

> TODO: Bytes for the dropped packets' payloads are estimated based on the average packet size. - What's about the other stats from this group? -> applicable for byteRcvLossTotal (as of now it has lost and out-of-order and undecrypted, but we do not size for lost and potentially out of order)

Reviewer: This should be mentioned in every stats which is based on the number of bytes of packets that have never arrived (even late).

Author: What's exactly mentioned in my TODO except you have other suggestions than byteRcvLossTotal.

### byteRcvUndecryptTotal

Same as `pktRcvUndecryptTotal`, but expressed in bytes, including payload and all the headers (20 bytes IPv4 + 8 bytes UDP + 16 bytes SRT). Receiver side.

## Interval-Based Statistics

### pktSent

Same as `pktSentTotal`, but for a specified interval. Available for sender.

### pktRecv

Same as `pktRecvTotal`, but for a specified interval. Available for receiver.

### pktSndLoss

Same as `pktSndLossTotal`, but for a specified interval. Available for sender.

### pktRcvLoss

Same as `pktRcvLossTotal`, but for a specified interval. Available for receiver.

### pktRetrans

Same as `pktRetransTotal`, but for a specified interval. Available for sender.

### pktRcvRetrans

Same as `pktRcvRetransTotal`, but for a specified interval.

TODO: 

> - There is no `pktRcvRetransTotal` stats.

Reviewer: Then this needs a fix.

> - Which side

Reviewer: Receiver. These are data packets that arrived with flag R=1.

TODO: Update accordingly.

### pktRecvACK

Same as `pktRecvACKTotal`, but for a specified interval.

TODO: Update accordingly.

### pktSentNAK

Same as `pktSentNAKTotal`, but for a specified interval.

TODO: Update accordingly.

### pktRecvNAK

Same as `pktRecvNAKTotal`, but for a specified interval.

TODO: Update accordingly.

### pktSndFilterExtra

Same as `pktSndFilterExtraTotal`, but for a specified interval. Sender only.

Introduced in v1.4.0. Refer to [SRT Packet Filtering & FEC](packet-filtering-and-fec.md).

### pktRcvFilterExtra

Same as `pktRcvFilterExtraTotal`, but for a specified interval. Receiver only.

Introduced in SRT v1.4.0. Refer to [SRT Packet Filtering & FEC](packet-filtering-and-fec.md).

### pktRcvFilterSupply

Same as `pktRcvFilterSupplyTotal`, but for a specified interval. Receiver only.

Introduced in SRT v1.4.0. Refer to [SRT Packet Filtering & FEC](packet-filtering-and-fec.md).

### pktRcvFilterLoss

Same as `pktRcvFilterLossTotal`, but for a specified interval. Receiver only.

Introduced in SRT v1.4.0. Refer to [SRT Packet Filtering & FEC](packet-filtering-and-fec.md).

### mbpsSendRate

Sending rate in Mbps. Sender side.

> TODO: How it is calculated?

Reviewer: `mbpsSendRate = 8 * pktSend / count_us(now - lastSampleTime)` (symbolically. `pktSend` is set to 0 at the `lastSampleTime` time).

### mbpsRecvRate

Receiving rate in Mbps. Receiver side.

TODO: How it is calculated?

### usSndDuration

Same as `usSndDurationTotal`, but measured on a specified interval. Available for sender.

### pktReorderDistance

> TODO: How it is calculated? Why it is interval based?

Reviewer: This is the maximum so far seen distance in sequence numbers between the out-of-order packet (late packet with R=0) and the top receiver sequence. This value is decreased by 1 together with reorder tolerance at the moment when 10 packets in a row were received in order.

The distance in sequence numbers between two original (not retransmitted) packets received out of order. Receiver only.

The traceable distance values are limited by the maximum reorder tolerance set by  `SRTO_LOSSMAXTTL`.

### pktReorderTolerance

TODO: 

> - Why it is in interval-based statistics?

Reviewer: Because reorder tolerance is self-regulating. It starts from 0 and it grows whenever `pktReorderDistance` grows, up to a maximum that is controlled by `SRTO_LOSSMAXTTL` option.

> 

Reviewer: Here is a little confusion in this description. There are two different stateous sequence numbers in the socket that are taken into account when regarding the incoming packet's sequence number:
- the top reception sequence
- the ACK sequence

Packets named "belated" are not late (or, not only) towards the top reception sequence, but they are late towards the ACK sequence. It means that these are packets that come too late to be regarded for anything. Packets that are "late", but not "belated", can still seal holes caused by missing packets, and usually these are retransmitted packets. Note that every belated packet means wasted link capacity.

It is important to note that for bandwidth estimation only data probing packets are taken into account while all the data packets (both data and data probing) are used for receiving speed estimation. The idea behind packet pair techniques is to send the groups of back-to-back packets, i.e., probing packet pairs, to a server thus making it possible to measure the minimum interval in receiving the consecutive packets.

### byteAvailSndBuf

The available space in the SRT sender buffer, in bytes. Sender only.

TODO: 

- increases first, decreases second or good?
- after the packets are sent over the UDP link - ?

This value decreases with the data scheduled for sending by the application and increases with every ACK received from the SRT receiver after the packets are sent over the UDP link.

### byteAvailRcvBuf

The available space in the SRT receiver buffer, in bytes. Receiver only.

> TODO: SRT socket?

Reviewer: UDP socket is not regarded anyhow in the stats anyway, but might be that this would be better for clarity.

The next received packet has sequence number 8. Reorder tolerance value is increased to 2.
The packet with sequence number 9 is reported lost.

### pktRcvAvgBelatedTime

> TODO: What's this?

Reviewer: This may need some attention. The intention of this data was to calculate the average time of a uselessly sent packet that has arrived despite that it was already received (or dropped) in order to get oriented how much the sender failed to prevent a potentially belated packet from being sent.

After time, I think in this form this stat may not exactly help in anything, as it tends to have a "negative cummulation" (packets that were retransmitted and accepted do not contribute to this value at all). OTOH it might give you an idea that there's something wrong if this value tends to be quite high and exceeds the latency (that is, it was sent later than 2* after the latency passed).

Accumulated difference between the current time and the time-to-play of a packet 
that is received late.

### pktRcvBelated

> TODO: Revise this, which side, measured over the interval

Reviewer: Receiver. This is the number of packets that had the sequence behind ACK and were sent uselessly.

Retransmitted packets can also be considered late.

### pktSndDrop

Same as `pktSndDropTotal`, but for a specified interval. Available for sender.

### pktRcvDrop

Same as `pktRcvDropTotal`, but for a specified interval. Available for receiver.

### pktRcvUndecrypt

Same as `pktRcvUndecryptTotal`, but for a specified interval. Available for receiver.

### byteSent

Same as `byteSentTotal`, but for a specified interval. Available for sender.

### byteRecv

Same as `byteRecvTotal`, but for a specified interval. Available for receiver.

### byteRcvLoss

Same as `byteRcvLossTotal`, but for a specified interval. Available for receiver.

### byteRetrans

Same as `byteRetransTotal`, but for a specified interval. Available for sender.

### byteSndDrop

Same as `byteSndDropTotal`, but for a specified interval. Available for sender.

### byteRcvDrop

Same as `byteRcvDropTotal`, but for a specified interval. Available for receiver.

### byteRcvUndecrypt

Same as `byteRcvUndecryptTotal`, but for a specified interval. Available for receiver.

## Instantaneous Statistics

### usPktSndPeriod

TODO: 

> - How is this calculated? The minimum time during which period?

Reviewer: This isn't calculated, it's the really used control value that represents the time interval between two consecutive sent packets. That is, when a sending thread is about to send a packet and the time since last sending is less than this, it sleeps up to that time.

Note that packets used for bandwidth probing and control packets are sent without respecting this interval, however (if I'm not missed) for bandwidth probing packets this haste time is recorded as spared time (this means longer sleeping for the next packet).

- probing packets
- rephrase - They may have different pacing of the outgoing packets, but all the packets will
  be placed in the same sending queue, which may affect the send timing. - who may have sockets? which may affect?

The current minimum packet inter-sending time, in microseconds. Does not take into account the probing packet pairs. Available for sender.

`usPktSndPeriod` is the minimum time (minimum sending period) that must be kept
between two packets sent consecutively over the link used by an SRT socket.
It is not the EXACT time interval between two consecutive packets. In the case where the time spent by an 
application between sending two consecutive packets exceeds `usPktSndPeriod`, the next 
packet will be sent faster, or even immediately, to preserve the average sending rate.

Note that several SRT sockets sharing one outgoing port use the same sending queue.
They may have different pacing of the outgoing packets, but all the packets will
be placed in the same sending queue, which may affect the send timing.

### pktFlowWindow

TODO:

- Rephrase this - The maximum number of packets that can be "in flight" state. - it does not reflect the idea

> - revise the whole paragraph

Reviewer: "In flight" means the packets that have departed the sender, but did not arrive at the receiver, in a certain point in time. Precise estimation of this value is impossible without having some 3rd independent probe machine that can retrieve data at precisely the same time from both machines simultaneously, but this value can at least exist in theory. The only point in time when this value can be calculated closely to reality is at the moment when ACK is being received by the data sender, and it's the distance between ACK and the top sender sequence - though this still includes the number of packets that have been received during the time when the ACK packet was travelling through the network.

In this value this is the "maximum" allowed flight size - because the sender can only control those packets that it didin't sent yet, and therefore if the measured flight size exceeds the maximum allowed flight size, it must stop sending, even though the receiver buffer free space didn't get to 0.

### pktCongestionWindow

The current congestion window size, in packets. Sender only.

The congestion window dynamically limits the maximum number of packets that can be "in flight" state. During transmission, congestion control module dynamically changes the size of congestion window.

In **file mode**, the size of congestion control window is equal to 16 packets at the very beginning and is increased during transmision to the number of reported acknowledged packets. This value is also updated based on the delivery rate reported by the receiver. It represents the maximum number of packets that can be safely sent without causing network congestion. The higher this value is, the faster the packets can be sent. In **live mode**, this field is not used.

### pktFlightSize

The number of packets in flight. Sender only.

The number of packets in flight is calculated as the difference between sequence numbers of the latest acknowledged packet (latest reported by an ACK message packet) and the latest sent packet at the moment statistic is being read. Note that `pktFlightSize <= pktFlowWindow` and `pktFlightSize <= pktCongestionWindow`.

> **NOTE:** ACKs are received by the SRT sender periodically at least every 10 milliseconds. This statistic is most accurate just after receiving an ACK packet and becomes a little exaggerated over time until the next ACK packet arrives. This is because with a new packet sent, while the ACK number stays the same for a moment, the value of `pktFlightSize` increases. But the exact number of packets arrived since the last ACK report is unknown. A new statistic might be added to only report the distance between the ACK sequence number and the sent packet sequence number at the moment when an ACK arrives. This statistic will not be updated until the next ACK packet arrives. The difference between the suggested statistic and `pktFlightSize` would then reveal the number of packets with an unknown state at that moment.

Reviewer: IMPORTANT: this "new statistic" mentioned here is now possible to be easily implemented because I had to introduce this "minimum flight window" tracer due to the need of implementing the "window" balancing algorithm for balancing groups.

### msRTT

The estimation for the round-trip time (RTT), in milliseconds. Available both for sender and receiver.

This value is calculated by the SRT receiver based on the incoming ACKACK control packets (sent back by the SRT sender to acknowledge incoming ACKs).

> TODO: peer to agent terminology, with the same journal

Reviewer: Instead of "peer to agent" etc. you can also say something like "from machine A to machine B and vice versa", though I'm not sure if this speaks better. The journal number is the data that is recorded when ACK message is sent so that it's known what the time was when sending ACK for which the received ACKACK message is a response.
